### PR TITLE
job controller changes

### PR DESCRIPTION
**Pull request by pdgetrf**
(1) added job controller changes to convert job spec resource update to annotation
 (2) fixed the duplicated update issue in (1)
 (3) added unit tests

tested end-to-end on a local environment

### CODE REVIEW

File: pkg/controller/job/job_controller.go [hunk 1]
```diff
@@ -400,6 +401,83 @@ func (jm *JobController) processNextWorkItem() bool {
...
+	cm := controller.NewPodControllerRefManager(jm.podControl, j, nil, controllerKind, nil)
+
+	cMap := make(map[string]*v1.Container)
+	for i, container := range j.Spec.Template.Spec.Containers {
+		cMap[container.Name] = &(j.Spec.Template.Spec.Containers[i])
+	}
+
+	for _, pod := range pods {
+		resourceUpdates := mergeResourceChanges(pod, cMap)
+		anno := make(map[string]string)
```
Comment (XiaoningDing) [inline pkg/controller/job/job_controller.go]: A minor one: better to move the line "anno := make(map[string]string)" into the if block

File: pkg/controller/job/job_controller.go [hunk 2]
```diff
@@ -400,6 +401,83 @@ func (jm *JobController) processNextWorkItem() bool {
...
+
+	// add annotation to pod if request and/or limit from job spec is different
+	for cid, podContainer := range pod.Spec.Containers {
+		hasUpdate := false
+		var patch v1.Container
+		jobSpecContainerResource := cMap[podContainer.Name].Resources
+
+		// update request from job spec
+		if !reflect.DeepEqual(podContainer.Resources.Requests, jobSpecContainerResource.Requests) {
+			if podContainer.Resources.Requests == nil {
```
Comment (XiaoningDing) [inline pkg/controller/job/job_controller.go:415]: It looks simpler to just return the entire "jobSpecContainerResource.Requests" if "podContainer.Resources.Requests" is nil. No need to make an empty map and then compare & add one by one.  The same applies to .Limits.

**Review by XiaoningDing — APPROVED** [review]

File: pkg/controller/job/job_controller.go [hunk 3]
```diff
@@ -400,6 +401,83 @@ func (jm *JobController) processNextWorkItem() bool {
...
+	for i, container := range j.Spec.Template.Spec.Containers {
+		cMap[container.Name] = &(j.Spec.Template.Spec.Containers[i])
+	}
+
+	for _, pod := range pods {
+		resourceUpdates := mergeResourceChanges(pod, cMap)
+		anno := make(map[string]string)
+		if len(resourceUpdates) > 0 {
+			jsonStr, _ := json.Marshal(resourceUpdates)
+			anno["vscale"] = string(jsonStr)
```
Comment (vinaykul) [inline pkg/controller/job/job_controller.go:467]: Let's remember to change this to pkg.scheduler.api.AnnotationResizeResources (part of my change)

File: pkg/controller/job/job_controller.go [hunk 4]
```diff
@@ -400,6 +401,83 @@ func (jm *JobController) processNextWorkItem() bool {
...
+
+			// only patch new annotation. ignore duplicate
+			if pod.Annotations == nil || pod.Annotations["vscale"] != anno["vscale"] {
+				cm.PatchPodResourceAnnotation(pod, anno)
+				glog.V(6).Infof("Adding resource update annotation %v to pod %s", anno, pod.Name)
+			} else {
+				glog.V(6).Infof("Ignord attempt to patch duplicated annotation %v to pod %s", anno, pod.Name)
```
Comment (vinaykul) [inline pkg/controller/job/job_controller.go:474]: Typo - 'Ignore'.
I'm wondering if this is causing the duplicate Update that I was seeing at the scheduler.

File: pkg/controller/job/job_controller.go (hunk 2 above)
Comment (pdgetrf) [reply pkg/controller/job/job_controller.go:415]: good catch! fixed.

also added tests for nil cases
